ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala

```diff
@@ -370,4 +351,33 @@ class InMemoryLedger(
       .getSource(startInclusive, None)
       .collect { case (offset, InMemoryConfigEntry(entry)) => offset -> entry }
 
+  override def deduplicateCommand(
+      deduplicationKey: String,
+      submittedAt: Instant,
+      ttl: Instant): Future[Option[CommandDeduplicationEntry]] =
+    Future.successful(this.synchronized {
+      commands
+        .get(deduplicationKey)
+        .fold[Option[CommandDeduplicationEntry]] {
+          commands += (deduplicationKey -> CommandDeduplicationEntry(
+            deduplicationKey,
+            submittedAt,
+            ttl,
+            None))
+          None
+        }(ce => Some(ce))
+    })
```

Comment:
Given that `getOrElseUpdate` seems to have the wrong semantics for your case, I would suggest we use an imperative style to highlight that we're causing side effects. Would this make it?
```scala
    Future.successful {
      this.synchronized {
        val entry = commands.get(deduplicationKey)
        if (!entry.isDefined) {
          commands += (deduplicationKey -> CommandDeduplicationEntry(
            deduplicationKey,
            submittedAt,
            ttl,
            None))
        }
        entry
      }
    }
```
Take it as a general suggestion and feel free to drop this here, though. After all, this will soon go away. :wink:

Reply:
Other thinking point/general observation: could we restrict the synchronization scope to the mutable state we're accessing (i.e. `commands.synchronized` instead of `this.synchronized`)?

Reply:
`scala.collection.concurrent.Map` has a `putIfAbsent` function which does this. We might consider using `TrieMap` rather than `mutable.Map`.

Reply:
@stefanobaghino-da 's suggestion sounds good. I will not further optimize this though, as this code will go away (as you noted).